### PR TITLE
Add test for checking outputs' consistency

### PR DESCRIPTION
**Pull request by Ostrokrzew**
Signed-off-by: Slawomir Jankowski <slawomir.jankowski@intel.com>

### CODE REVIEW

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        core_part = core_dev.partitions[0]
+        blocks_in_cache = int(cache_size / cache_line_size.value)
+
+    with TestRun.step("Start cache and add core with filesystem."):
+        cache = start_cache(cache_part, cache_mode, cache_line_size, force=True)
+        core_part.create_filesystem(Filesystem.xfs)
+        exp_obj = cache.add_core(core_part)
+
+    for _ in TestRun.iteration(range(iterations), f"Run configuration {iterations} times"):
+        with TestRun.step("Reset stats and put workload on exported object."):
```
Comment (katlapinka) [inline test/functional/tests/stats/test_consistency_between_outputs.py:43]: run workload?

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            cache.reset_counters()
+            random_count = random.randint(blocks_in_cache / 32, blocks_in_cache / 2)
```
Comment (katlapinka) [inline test/functional/tests/stats/test_consistency_between_outputs.py:45]: I'm just curious - why divided by 32 and by 2? Maybe add some comment

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        core_dev.create_partitions([cache_size * 4])
+        core_part = core_dev.partitions[0]
+        blocks_in_cache = int(cache_size / cache_line_size.value)
+
+    with TestRun.step("Start cache and add core with filesystem."):
+        cache = start_cache(cache_part, cache_mode, cache_line_size, force=True)
+        core_part.create_filesystem(Filesystem.xfs)
+        exp_obj = cache.add_core(core_part)
+
+    for _ in TestRun.iteration(range(iterations), f"Do next steps for {iterations} times"):
```
Comment (katlapinka) [inline test/functional/tests/stats/test_consistency_between_outputs.py:78]: "Repeat next steps {iterations} times"  - without "for"

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        with TestRun.step("Reset stats and put workload on exported object."):
+            exp_obj.reset_counters()
+            random_count = random.randint(blocks_in_cache / 32, blocks_in_cache / 2)
```
Comment (katlapinka) [inline test/functional/tests/stats/test_consistency_between_outputs.py:81]: same as in above test

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+iterations = 64
+cache_size = Size(8, Unit.GibiByte)
+
+
+@pytest.mark.parametrize("cache_line_size", CacheLineSize)
+@pytest.mark.parametrize("cache_mode", CacheMode.with_any_trait(CacheModeTrait.InsertRead
+                                                                | CacheModeTrait.InsertWrite))
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_cache_output_consistency(cache_line_size, cache_mode):
```
Comment (katlapinka) [inline test/functional/tests/stats/test_consistency_between_outputs.py:27]: add test description

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+            if csv_stats != table_stats:
+                TestRun.LOGGER.error(f"Inconsistent outputs:\n{csv_stats}\n\n{table_stats}")
+
+
+@pytest.mark.parametrize("cache_line_size", CacheLineSize)
+@pytest.mark.parametrize("cache_mode", CacheMode.with_any_trait(CacheModeTrait.InsertRead
+                                                                | CacheModeTrait.InsertWrite))
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_core_output_consistency(cache_line_size, cache_mode):
```
Comment (katlapinka) [inline test/functional/tests/stats/test_consistency_between_outputs.py:63]: add test description

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        else:
+            part_of_keys, part_of_values = parse_tables_section(output_parts[i])
+
+        # receive keys and values lists from every section
+        keys.extend(part_of_keys)
+        values.extend(part_of_values)
+
+    # put the keys and the values into a dictionary
+    dictionary_output = {}
+    for i in range(len(keys)):
```
Comment (mmichal10) [inline test/functional/tests/stats/test_consistency_between_outputs.py:175]: `for key, value in zip(keys, values):` seems more pythonic

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    for i in range(len(output_parts)):
+        # the first part is config section
+        if i == 0:
+            part_of_keys, part_of_values = (
+                parse_core_conf_section(output_parts[i]) if core_id
+                else parse_cache_conf_section(output_parts[i])
+            )
+        # the remaining parts are table sections
+        else:
+            part_of_keys, part_of_values = parse_tables_section(output_parts[i])
```
Comment (mmichal10) [inline test/functional/tests/stats/test_consistency_between_outputs.py:167]: You can move parsing conf section before loop, remove first value from `output_parts` and use `for section in output_parts:` which seems more pythonic.

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+        values.append(splitted_line[3])
+
+    return keys, values
+
+
+def is_table_separator(line):
+    """
+    Tables in the 'casadm -P' output have plus signs only on separator lines.
+    """
+    return '+' in line
```
Comment (Deixx) [inline test/functional/tests/stats/test_consistency_between_outputs.py:277]: this is not always true - depending on OS settings, tables can have different styles:
```
╔══════════════════╤═════════╤═══════╤═════════════╗
║ Usage statistics │  Count  │   %   │   Units     ║
╠══════════════════╪═════════╪═══════╪═════════════╣
║ Occupancy        │     524 │   0.0 │ 4KiB Blocks ║
║ Free             │ 1297172 │ 100.0 │ 4KiB Blocks ║
║ Clean            │     524 │   0.0 │ 4KiB Blocks ║
║ Dirty            │       0 │   0.0 │ 4KiB Blocks ║
╚══════════════════╧═════════╧═══════╧═════════════╝
```
return '|' not in line # might address that, although I'm not sure if vertical line in this style is the same as the keyboard char

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+
+
+def dd_builder(cache_mode, cache_line_size, count, device):
+    dd = (Dd()
+          .block_size(cache_line_size.value)
+          .count(count))
+
+    if CacheModeTrait.InsertRead in CacheMode.get_traits(cache_mode):
+        dd.input(device.system_path).output("/dev/null").iflag("direct")
+
```
Comment (Deixx) [inline test/functional/tests/stats/test_consistency_between_outputs.py:287]: why empty line before else? suggests separate block of code

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+        table_as_list.pop(0)
+
+    # remove separator lines, it is also useless
+    for line in table_as_list:
+        if is_table_separator(line):
+            table_as_list.remove(line)
+
+    # split lines to columns by '|' sign and remove whitespaces
+    for line in table_as_list:
+        splitted_line = line.split('|')
```
Comment (Deixx) [inline test/functional/tests/stats/test_consistency_between_outputs.py:256]: see my comment for line 277

File: test/functional/tests/stats/test_consistency_between_outputs.py (hunk 9 above)
Comment (Ostrokrzew) [reply test/functional/tests/stats/test_consistency_between_outputs.py]: Unfortunately, it is not. However I found solution, thanks to Your feedback.

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+from core.test_run import TestRun
+from storage_devices.disk import DiskType, DiskTypeSet, DiskTypeLowerThan
+from test_tools.dd import Dd
+from test_tools.disk_utils import Filesystem
+from test_utils.size import Size, Unit
+
+iterations = 64
+cache_size = Size(8, Unit.GibiByte)
+
+
```
[the remark this answers is not in the document]
Comment (Ostrokrzew) [reply test/functional/tests/stats/test_consistency_between_outputs.py:21]: good point

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+@pytest.mark.parametrize("test_object", ["cache", "core"])
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_output_consistency(cache_line_size, cache_mode, test_object):
+    """
+        title: Test consistency between different cache statistics' outputs.
+        description: |
+          Check if OpenCAS's statistics for cache are consistent regardless of the output format.
+        pass_criteria:
+          - Cache statistics in CSV format matches cache statistics in table format.
```
Comment (katlapinka) [inline test/functional/tests/stats/test_consistency_between_outputs.py:34]: update description - tests are now merged

Comment (Ostrokrzew) [reply test/functional/tests/stats/test_consistency_between_outputs.py]: Thx, I forgot to update it.

File: test/functional/tests/stats/test_consistency_between_outputs.py [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_output_consistency(cache_line_size, cache_mode, test_object):
+    """
+        title: Test consistency between different cache and core statistics' outputs.
+        description: |
+          Check if OpenCAS's statistics for cache and core are consistent
+          regardless of the output format.
+        pass_criteria:
+          - Cache statistics in CSV format matches cache and core statistics in table format.
```
Comment (katlapinka) [inline test/functional/tests/stats/test_consistency_between_outputs.py:35]: "Statistics in CSV(...)" - not cache statistics

**Review by katlapinka — APPROVED** [review]

**Review by Deixx — APPROVED** [review]

**Pull request merged**